### PR TITLE
Fix_wxListCtrl_HitTest

### PR DESCRIPTION
**Pull request by catalinr**
This should also fix trac ticket 17059.

### CODE REVIEW

File: src/msw/listctrl.cpp [hunk 1]
```diff
@@ -1780,37 +1774,41 @@ wxListCtrl::HitTest(const wxPoint& point, int& flags, long *ptrSubItem) const
...
     flags = 0;
 
-    if ( hitTestInfo.flags & LVHT_ABOVE )
-        flags |= wxLIST_HITTEST_ABOVE;
-    if ( hitTestInfo.flags & LVHT_BELOW )
-        flags |= wxLIST_HITTEST_BELOW;
+    // test whether an actual item was hit or not
+    if ( hitTestInfo.flags == LVHT_ONITEMICON )
+        flags = wxLIST_HITTEST_ONITEMICON;
+    // note a bug in comtl32.dll:
```
Comment (vadz) [inline src/msw/listctrl.cpp]: Typo: `comctl32.dll`.

File: src/msw/listctrl.cpp [hunk 2]
```diff
@@ -1780,37 +1774,41 @@ wxListCtrl::HitTest(const wxPoint& point, int& flags, long *ptrSubItem) const
...
+    // 1) in report mode, when there are 2 or more columns, if you click to the
+    //    right of the first column, ListView_HitTest() returns LVHT_ONITEM
+    //    (LVHT_ONITEMICON|LVHT_ONITEMLABEL|LVHT_ONITEMSTATEICON).
+    else if ( hitTestInfo.flags == LVHT_ONITEMLABEL ||
```
Comment (vadz) [inline src/msw/listctrl.cpp]: Couldn't/shouldn't we just test for `flags & LVHT_ONITEMLABEL`? I.e. are there any cases when `LVHT_ONITEMLABEL` is set, but we don't want to return `wxLIST_HITTEST_ONITEMLABEL`?

File: src/msw/listctrl.cpp [hunk 3]
```diff
@@ -1780,37 +1774,41 @@ wxListCtrl::HitTest(const wxPoint& point, int& flags, long *ptrSubItem) const
...
+              (hitTestInfo.flags == LVHT_ONITEM && InReportView()) )
+        flags = wxLIST_HITTEST_ONITEMLABEL;
```
Comment (vadz) [inline src/msw/listctrl.cpp]: This cascaded `if` would be more readable if we put braces around the statement bodies, especially this one where the condition itself is on multiple lines (but then all the other ones too, for consistency).

File: tests/controls/listbasetest.cpp [hunk 4]
```diff
@@ -445,6 +445,63 @@ void ListBaseTestCase::ImageList()
...
+
+    // label, beyond column 0
+    wxRect rectItem;
+    list->GetItemRect(0, rectItem); // entire item
+    int xHit = rectSubItem0.GetRight() + (rectItem.GetRight() - rectSubItem0.GetRight()) / 2;
+    list->HitTest(wxPoint(xHit, y), flags);
+    CPPUNIT_ASSERT_EQUAL_MESSAGE("Expected wxLIST_HITTEST_ONITEMLABEL",
+        wxLIST_HITTEST_ONITEMLABEL, flags);
+
+    //tidy up when we are finished
```
Comment (vadz) [inline tests/controls/listbasetest.cpp]: I see that this was already the case before your changes, but I don't understand why do we need to tidy up here when the list is going to be deleted anyhow in `tearDown()`. I.e. I'd rather remove the existing calls to `ClearAll()` than add yet another one.

**Review by vadz — COMMENTED** [review]
Globally this looks good, thanks!

I am still not sure whether the new test can/should be enabled for non-MSW ports, i.e. does it pass with wxGTK too? If not, we should put it inside `#ifdef __WXMSW__` checks to at least avoid breaking the test suite there -- even if making the test pass would be much better, of course.

**Comment by a-wi** [comment]
I have pushed all commits (squashed a bit, as 640b7df6, c4c7f129, 2c7435a4, 6c9b3a82, fb0a1182) except 32b6cb9 because this one refers to the issue [17059](http://trac.wxwidgets.org/ticket/17059) which would need more investigation, I think. It seems that wxListCtrl::UpdateStyle() and SetWindowStyleFlag() pair should be examined because apparently they don't play well together.

**Comment by catalinr** [comment]
IIRC the main purpose of this PR was to fix [17059](http://trac.wxwidgets.org/ticket/17059). What exactly would need to be investigated further?
IMHO the only thing still pending is whether to enable the test case for `wxGTK` or not.

**Comment by a-wi** [comment]
Based on the title I would bet this PR is about HitTest(), not about UpdateStyle() :)

Honestly, your commit regarding UpdateStyle() looks to me rather as a workaround than a fix for [17059](http://trac.wxwidgets.org/ticket/17059). Apparently, intended purpose of UpdateStyle() is/was to synchronize between physical state of the control with its logical state (given by flag). Problem is that this synchronization is done already in SetWindowStyleFlag() (through MSWUpdateStyle()) which is called before and effectively UpdateStyle() does nothing (what was observed by OP). The question is what to do with UpdateStyle()? Is it necessary? Maybe it should be modified as it was suggested by OP?

FYI: Test under GTK fails, because generic implementation of HitTest() returns only wxLIST_HITTEST_ONITEMICON and wxLIST_HITTEST_ONITEMLABEL, not e.g wxLIST_HITTEST_ONITEMSTATEICON.

../../../tests/controls/listbasetest.cpp:473:Assertion
Test name: ListCtrlTestCase::HitTest
equality assertion failed
- Expected: 512
- Actual  : 32
- Expected wxLIST_HITTEST_ONITEMSTATEICON

**Pull request closed without merging**